sycl/doc/Assert.md

```diff
@@ -0,0 +1,269 @@
+variable. This allows to refer to the accessor without modifying each and every
+user's function. Fallback implementation of `__devicelib_assert_fail` restores
+accessor metadata from program scope variable and writes assert information to
+the assert buffer. Atomic operations are used in order to not overwrite existing
+information.
+
+DPCPP Runtime checks contents of the assert buffer for assert failure flag after
+kernel finishes.
+
+Both storing of accessor metadata and writing assert failure is performed with
+help of built-ins. Implementations of these builtins are substituted by
+frontend.
+
+#### Built-ins operation
+
+Accessor is a pointer augmented with offset and two ranges (access range and
+memory range).
+
+There are two built-ins provided by frontend:
+ * `__store_acc()` - to store accessor metadata into program-scope variable.
+ * `__store_assert_failure()` - to store flag about assert failure in a buffer
+   using the metadata stored in program-scope variable.
+
+The accessor should be stored to program scope variable in global address space
+using atomic operations. Motivation for using atomic operations: the program may
+contain several kernels and some of them could be running simultaneously on a
+single device.
+
+The `__store_assert_failure()` built-in atomically sets a flag in a buffer. The
+buffer is accessed using accessor metadata from program-scope variable. This
+built-in return a boolean value which is `true` if the flag is set by this call
+to `__store_assert_failure()` and `false` if the flag was already set.
+Motivation for using atomic operation is the same as with `__store_acc()`
+builtin.
+
+The following pseudo-code snippets shows how these built-ins are used.
+First of all, assume the following code as user's one:
+```
+void user_func(int X) {
+  assert(X && “X is nil”);
+}
+
+int main() {
+  queue Q(...);
+  Q.submit([&] (handler& CGH) {
+    CGH.single_task([=] () {
+      do_smth();
+      user_func(0);
+      do_smth_else();
+    });
+  });
+  ...
+}
+```
+
+The following LLVM IR pseudo code will be generated for the user's code:
+```
+@AssertBufferPtr = global void* null
+@AssertBufferAccessRange = ...
+@AssertBufferMemoryRange = ...
+@AssertBufferOffset = ...
+
+/// user's code
+void user_func(int X) {
+if (!(X && “X is nil")) {
+    __assert_fail(...);
+  }
+}
+
+users_kernel(...) {
+  do_smth()
+  user_func(0);
+  do_smth_else();
+}
+
+/// a wrapped user's kernel
+kernel(AssertBufferAccessor, OtherArguments...) {
+  __store_acc(AssertBufferAccessor);
+  users_kernel(OtherArguments...);
+}
+
+/// __assert_fail belongs to Linux version of devicelib
+void __assert_fail(...) {
+  ...
+  __devicelib_assert_fail(...);
+}
+
+void __devicelib_assert_fail(Expr, File, Line, GlobalID, LocalID) {
+  ...
+  if (__store_assert_info())
+    printf("Assertion `%s' failed in %s at line %i. GlobalID: %i, LocalID: %i",
+           Expr, File, Line, GlobalID, LocalID);
```

Comment:
This may be a silly question, but do we have device side printf()?  I would have expected this string to be written to the assert buffer, so the DPC++ runtime could retrieve and print it.

Reply:
Yes, we do have a device-side `printf`.
Can't see use-case when fetching the message is required.
Does it look inconsistent with the assert exception throw?

Reply:
> Does it look inconsistent with the assert exception throw?

This is an interesting question.  What string does the exception contain?  I.e. what does the application get if it calls `e.what()` on that exception?

Reply:
From my point of view a generic message like `Assert had happened` will be enough for return value of `e.what()`.
For the time being, `assert` prints the following to `stderr`:
```
<file>:<line>: <function>: global id: [%lu,%lu,%lu], local id: [%lu,%lu,%lu] Assertion `<expr>` failed.
```

Reply:
Added to doc.